packages/react-bloc/test/bloc-provider.test.tsx

```diff
@@ -0,0 +1,202 @@
+import { BlocProvider, BlocBuilder, BlocConsumer } from '../lib/react-bloc'
+import * as React from 'react'
+import { Bloc } from '@felangel/bloc'
+import { mount } from 'enzyme'
+import 'jsdom-global/register'
+
+enum CounterEvent {
+  increment,
+  decrement
+}
+
+class CounterBloc extends Bloc<CounterEvent, number> {
+  constructor() {
+    super(0)
+  }
+
+  async *mapEventToState(event: CounterEvent): AsyncIterableIterator<number> {
+    switch (event) {
+      case CounterEvent.increment:
+        yield this.state + 1
+        break
+      case CounterEvent.decrement:
+        yield this.state - 1
+        break
+    }
+  }
+}
+
+type CounterAppProps = {
+  title: string
+}
+
+class CounterComponent extends React.Component<{}, any> {
+  render() {
+    return (
+      <React.Fragment>
+        <BlocBuilder<CounterBloc, number>
+          type="CounterBloc"
+          builder={count => {
+            return <p id="counter-with-type-lookup">{count}</p>
+          }}
+        />
+        <BlocConsumer
+          type="CounterBloc"
+          consumer={bloc => (
+            <React.Fragment>
+              <BlocBuilder<CounterBloc, number>
+                bloc={bloc}
+                builder={count => {
+                  return <p id="counter-without-type-lookup">{count}</p>
+                }}
+              />
+              <button
+                id="counter-increment"
+                onClick={() => {
+                  bloc.add(CounterEvent.increment)
+                }}
+              ></button>
+            </React.Fragment>
+          )}
+        />
+      </React.Fragment>
+    )
+  }
+}
+
+const CounterApp: React.FunctionComponent<CounterAppProps> = (
+  props: CounterAppProps
+): JSX.Element => {
+  const [type, setType] = React.useState('CounterBloc')
```

Comment:
I'm a bit confused about the separation of the bloc and the type. Can we not obtain the name of the bloc through something like `instance.constructor.name or Class.name` depending on if you are using the class or the instance to obtain the name. Inferring that the name of the bloc is the type is the case in the dart version i believe.

Reply:
Let me know if there is another reason behind this choice. Maybe I'm missing something.

Reply:
The constructor name is not reliable in JS as i mentioned in the other comment. It is almost always modified by uglifiers, minifiers, is not supported on old browsers and can fail for class inheritance in ES5.

https://stackoverflow.com/questions/10314338/get-name-of-object-or-class

The above link covers most of the issues.

Reply:
Also, I used "type" as the variable name, but something like "key" or "name" also works. As long as it is a unique id for the bloc instance.